src/web100srv.c

```diff
@@ -1527,647 +1326,364 @@ void child_process(int parent_pipe, SSL_CTX *ssl_context, int ctlsockfd) {
+  new_child->pipe = child_pipe[1];
+  new_child->running = 0;
+  new_child->next = NULL;
+
+  return new_child;
+}
 
-    FD_ZERO(&rfd);
-    FD_SET(listenfd, &rfd);
-    if (waiting > 0) {
-      // there are clients waiting
-      sel_tv.tv_sec = 3;  // 3 seconds == WAIT_TIME_SRVR
-      sel_tv.tv_usec = 0;
-      log_println(3, "Waiting for new connection, timer running");
-      do {
-        retcode = select(listenfd + 1, &rfd, NULL, NULL, &sel_tv);
-        // retry if a signal caused select() to exit.
-      } while ((retcode == -1) && (errno == EINTR));
-      tt = time(0);
+/**
+ * Enqueues a new ndtchild struct onto the queue of clients to be run.  Will
+ * not enqueue the child unless the server has spare capacity.  This function
+ * takes ownership of the memory pointed to by new_child, and either frees that
+ * memory or passes ownership to the linked list that goes from head to tail.
+ * @param new_child The new client
+ * @param head The head of the client queue
+ * @param tail The tail of the client queue
+ */
+void enqueue_new_child(struct ndtchild *new_child, struct ndtchild **head,
+                       struct ndtchild **tail) {
+  ParentChildSignal child_signal;
+  struct ndtchild *tmp_ptr;
+  int client_count = 0;
+  // Count the number of connected clients
+  tmp_ptr = *head;
+  while (tmp_ptr != NULL) {
+    client_count++;
+    tmp_ptr = tmp_ptr->next;
+  }
 
+  // Check whether the server is overloaded
+  if ((!queue && client_count >= 1) ||
+      (queue && !multiple && client_count >= max_clients) ||
+      (queue && multiple && client_count >= 4 * max_clients)) {
+    // The server is overloaded. Reject the client.
+    log_println(0,
+                "Too many clients/mclients (%d) waiting to be served, "
+                "Please try again later.",
+                new_child->pid);
+    child_signal = SRV_QUEUE_SERVER_BUSY;
+    write(new_child->pipe, &child_signal, sizeof(ParentChildSignal));
+    // Whether the write succeeds or not, we are done with this child. It
+    // will either kill itself in response to this message, or it will kill
+    // itself with its own watchdog timer. Either way, not our problem
+    // anymore. Send the message, close the pipe, and move on.
+    close(new_child->pipe);
+    log_println(6, "Too many clients, freeing child=0x%x", new_child);
+    free(new_child);
+    new_child = NULL;
+  } else {
+    // The server is not overloaded, so queue up the new client
+    if (*head == NULL) {
+      *head = *tail = new_child;
     } else {
-      // Nothing is in the queue, so wait forever until a new connection
-      // request arrives
-      log_println(3, "Timer not running, waiting for new connection");
-      mclients = 0;
-      do {
-        retcode = select(listenfd + 1, &rfd, NULL, NULL, NULL);
-        // retry if a signal caused the select() to exit
-      } while ((retcode == -1) && (errno == EINTR));
+      (*tail)->next = new_child;
+      *tail = new_child;
     }
+  }
```

Comment:
Sold.  Made `wait_for_wakeup` return true if there was a new client waiting and false otherwise.

```
    // Wait for a new connection, an interruption, or a timeout.
    if (wait_for_wakeup(listenfd, signalfd, (head_ptr == NULL))) {
      new_child = spawn_new_child(listenfd, ssl_context);
      if (new_child != NULL) attempt_enqueue(new_child, &head_ptr);
    }
    // Perform queue maintenance: send messages to clients and reap the dead.
    perform_queue_maintenance(&head_ptr);
```